draft-ietf-tls-esni.md

```diff
@@ -255,97 +285,93 @@ extension, defined as follows:
+If the server accepts ECHO, it does not send this extension.
+If it rejects ECHO, then it sends the following structure in
 EncryptedExtensions:
 
 ~~~
    struct {
-       ESNIConfig retry_keys<1..2^16-1>;
+       ECHOConfigs retry_configs;
    } ServerEncryptedCH;
 ~~~
 
-retry_keys
-: One or more ESNIConfig structures containing the keys that the client should use on
-subsequent connections to encrypt the ClientESNIInner structure.
+retry_configs
+: An ESNIConfigs structure containing one or more ECHOConfig structures in
+decreasing order of preference that the client should use on subsequent
+connections to encrypt the ClientHelloInner structure.
 
-This protocol also defines the "esni_required" alert, which is sent by the
+This protocol also defines the "echo_required" alert, which is sent by the
 client when it offered an "encrypted_server_name" extension which was not
 accepted by the server.
 
+# The "echo_nonce" extension {#echo-nonce}
 
-# The "esni_nonce" extension {#esni-nonce}
-
-When using ESNI, the client MUST also add an extension of type
-"esni_nonce" to the ClientHelloInner (but not to the outer
-ClientHello).
-
-The encrypted SNI is carried in an "encrypted_server_name"
-extension, defined as follows:
+When using ECHO, the client MUST also add an extension of type
+"echo_nonce" to the ClientHelloInner (but not to the outer
+ClientHello). This extension is defined as follows:
 
 ~~~
    enum {
-       esni_nonce(0xffce), (65535)
+       echo_nonce(0xffce), (65535)
    } ExtensionType;
 
    struct {
        uint8 nonce[16];
-   } ESNINonce;
+   } ECHONonce;
 ~~~~
 
 nonce
-: A random 16-octet value generated by the client and echoed by the
-server.
+: A 16-byte nonce exported from the HPKE encryption context. See {{send-echo}}
+for details about its computation.
 
 Finally, requirements in {{client-behavior}} and {{server-behavior}} require
 implementations to track, alongside each PSK established by a previous
 connection, whether the connection negotiated this extension with the
-"esni_accept" response type. If so, this is referred to as an "ESNI PSK".
-Otherwise, it is a "non-ESNI PSK". This may be implemented by adding a new field
+"echo_accept" response type. If so, this is referred to as an "ECHO PSK".
+Otherwise, it is a "non-ECHO PSK". This may be implemented by adding a new field
 to client and server session states.
 
 ## Incorporating Outer Extensions {#outer-extensions}
 
 Some TLS 1.3 extensions can be quite large
-and having them both in the inner and outer ClientHello wil lead to
+and having them both in the inner and outer ClientHello will lead to
 a very large overall size. One particularly pathological example
 is "key_share" with post-quantum algorithms. In order to reduce
 the impact of duplicated extensions, the client may use the
 "outer_extension" extension.
 
 ~~~
    enum {
-       esni_extension(TBD), (65535)
+       outer_extension(TBD), (65535)
```

Comment:
This does force the server to implement somewhat fussy reconstruction logic, but I guess we can fiddle with this depending on how fussy it ends up being.

One comment though: rather than a hash per extension, why not include a single hash of the final ClientHello? That way the marginal cost of compressing an extension is much smaller.

Reply:
Agreed -- we can muck with this as needed. Offline discussions with @martinthomson led to something similar (replace with a hash for all substituted extensions), though a hash for the entire reconstructed CH would also work.

Reply:
I made this suggestion. @martinthomson @davidben, please have a look!